### PR TITLE
Feature: Domstream screen resolution

### PR DESCRIPTION
**Pull request by Maaiins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tested the changes
- [x] Build (`/path/to/php cli.php cmd=build`) was run locally and any changes were pushed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Other (please describe): 

## What is the current behavior?
Domstreams are rendered on the current window resolution

Issue Number: #301 

## What is the new behavior?
Domstreams are rendered in an separated iframe now, which respects the user screen resolution.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

Database migration needed for fiel `page_width` (BIGINT) and `page_height` (BIGINT) in `owa_domstream`. @padams due to lack of knowledge I didn´t do this for now feel free to checkout this branch and push the migration.

## Docs need to be updated?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by padams** [comment]
Here's how to create a schema updater:

1. create a schema updater class in `modules/base/updates/` named `owa_base_010_update` in a file called `010.php`. schema updates are applied (or un-applied) in sequential order by module. So in the case of the base module the next schema version would be 10 - hence the naming convention.

1.  Schema update classes need to define two methods:
    - `up()` - called when the existing install needs to apply this update
       - In the case of this update you are just adding columns so use the `addColumn` method in `owa_entity`. see update class 006.php for an example.
    - `down()` - called when the install needs to go back to the prior version of the schema. Think of this method as a revert to prior schema state method. I've never needed to use this but it's good practice just in case...

3. Tell the base module that it now requires a new version of the schema. This is done by changing  the `$this->version` and `$this->required_schema_version` variables in the module's `module.php` constructor.

4. Trigger the update by either:
    - navigating to any page of the admin interface which will get intercepted to show an upgrade button
    - or run `update` cmd via the cli.

**Comment by Maaiins** [comment]
Don´t merge this one at the moment!

It's A bit confusing, the domstream entity and logic is inside the base module but there's another domstream module. Shouldn't that all in the domstream module? Or is there any reason?

Now the real problem i created the files but the admin page is blank... There´s the log:
```
15:03:35 2020-04-15 3449 [debug_log] check for http host 
15:03:35 2020-04-15 3449 [debug_log] Setting cookie domain to .xxx 
15:03:35 2020-04-15 3449 [debug_log] *** Starting Open Web Analytics v1.6.7. Running under PHP v7.2.24-nmm1 (Linux) *** 
15:03:35 2020-04-15 3449 [debug_log] Request URL:GET /owa/index.php 
15:03:35 2020-04-15 3449 [debug_log] User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.163 Safari/537.36 
15:03:35 2020-04-15 3449 [debug_log] Host: xxx 
15:03:35 2020-04-15 3449 [debug_log] check cache file: /owa/owa-data/caches/1/owa_configuration/fafe1b60c24107ccd8f4562213e44849.php 
15:03:35 2020-04-15 3449 [debug_log] CACHE HIT (Cold) - Retrieved Object from Cache File - Collection: owa_configuration, id: fafe1b60c24107ccd8f4562213e44849 
15:03:35 2020-04-15 3449 [debug_log] Caller configuration overrides applied. 
15:03:35 2020-04-15 3449 [debug_log] Filtering register_build_package 
15:03:35 2020-04-15 3449 [debug_log] Filtering register_build_package 
15:03:35 2020-04-15 3449 [debug_log] Filtering register_build_package 
15:03:35 2020-04-15 3449 [debug_log] base Schema version is 9 
15:03:35 2020-04-15 3449 [debug_log] base Required Schema version is 10 
15:03:35 2020-04-15 3449 [debug_log] domstream Schema version is 1 
15:03:35 2020-04-15 3449 [debug_log] domstream Required Schema version is 1 
15:03:35 2020-04-15 3449 [debug_log] Getting state - store: u key:  
15:03:35 2020-04-15 3449 [debug_log] loading last value in initial state container for store: u 
15:03:35 2020-04-15 3449 [debug_log] populating state for store: u, name: , value: admin, store type: cookie, is_perm:  
15:03:35 2020-04-15 3449 [debug_log] Notifying listeners of init 
15:03:35 2020-04-15 3449 [debug_log] no listeners registered for this event type. 
15:03:35 2020-04-15 3449 [debug_log] EHS: Responses - Array
(
)
 
15:03:35 2020-04-15 3449 [debug_log] EHS: Event was handled successfully by all handlers. 
15:03:35 2020-04-15 3449 [debug_log] Handling request with params: Array
(
)
 
15:03:35 2020-04-15 3449 [debug_log] About to perform action: base.sites 
15:03:35 2020-04-15 3449 [debug_log] Performing Action: owa_sitesController 
15:03:35 2020-04-15 3449 [debug_log] Cannot perform action. OWA Updates required. 
15:03:35 2020-04-15 3449 [debug_log] Total session time: 0.0195858478546 
15:03:35 2020-04-15 3449 [debug_log] goodbye from OWA 
15:03:35 2020-04-15 3449 [debug_log] starting to persist cache... 
15:03:35 2020-04-15 3449 [debug_log] There seem to be no dirty objects in the cache to persist. 
15:03:35 2020-04-15 3449 [debug_log] Cache Statistics: 
                          Total Hits: 1 (Warm/Cold: 0/1)
                          Total Miss: 0
                          Total Added to Cache: 0
                          Total Replaced: 0
                          Total Persisted: 0
                          Total Removed: 0 
```

**Comment by padams** [comment]
thats a bug i introduced in my REST refactoring. I just pushed a commit in your branch to fix it.

**Comment by padams** [comment]
the domstream schema/table/entity started out life in the base module. I agree it should all be  moved to the domstream module. in that case you would need to:

 -  increment the schema version in the domstream module not the base module
 -  and relocate the domstream entity class into the domstream module directory
 -  register the entity in the domstream `module.php`
 - then search for uses of `base.domstream` by the `entityFactory` method and replace them with `domstream.domstream`

**Comment by Maaiins** [comment]
> the domstream schema/table/entity started out life in the base module. I agree it should all be moved to the domstream module. in that case you would need to:
> 
> * increment the schema version in the domstream module not the base module
> * and relocate the domstream entity class into the domstream module directory
> * register the entity in the domstream `module.php`
> * then search for uses of `base.domstream` by the `entityFactory` method and replace them with `domstream.domstream`

Isn´t it better to open an pull request for the refactoring?

**Comment by Maaiins** [comment]
> thats a bug i introduced in my REST refactoring. I just pushed a commit in your branch to fix it.

Nice thanks, i will checkout your change and test it first

**Comment by Maaiins** [comment]
BTW thanks for your activity. It´s very pleasant to develop such a project together unfortunately i never had that experience ;-)

**Comment by padams** [comment]
You could but don't you need the schema update to complete this one? If you apply the schema update in the base and then move the entity to the domstream module it might get confusing...

**Comment by padams** [comment]
@Maaiins unrelated but send me an email when you get a chance.

**Comment by Maaiins** [comment]
> You could but don't you need the schema update to complete this one? If you apply the schema update in the base and then move the entity to the domstream module it might get confusing...

True your right, never thought about that. Will work on that.

File: modules/base/updates/010.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        $domstreams_columns = [
+            'page_width',
+            'page_height',
+        ];
+
+        // add columns to owa_domstream
+        foreach ( $domstreams_columns as $domstreams_column ) {
+            $ret = $domstream->dropColumn( $domstreams_column );
+            if ( $ret === true ) {
+                $this->e->notice( "$domstreams_column added to owa_domstream" );
```
Comment (padams) [inline modules/base/updates/010.php:72]: Looks like you need ot change these notices to say the columns were "removed".

File: modules/base/updates/010.php [hunk 2]
```diff
@@ -32,7 +32,7 @@
 class owa_base_010_update extends owa_update {
 
     var $schema_version = 10;
-    var $is_cli_mode_required = true;
+    var $is_cli_mode_required = false;
```
Comment (padams) [inline modules/base/updates/010.php:35]: worried about the browser request timeout before completing?

Comment (Maaiins) [reply modules/base/updates/010.php:35]: No, was just a copy paste mistake. And i think adding two columns wouldn´t result in timeout

**Comment by Maaiins** [comment]
Found some trouble, by moving domstream to a module.

There are old updates in base for domstream and the tracking logic is in the owa tracker.js.

I think we should ad an separated logic to add something to the tracker, before we move domstream out of base or is there any way?

**Comment by padams** [comment]
Hmm. I need to review the tracker code to see how feasible that is. I’d say for now let’s keep downstream entity in the base module and apply the update there. That way you can land this.

**Comment by Maaiins** [comment]
So i think we should merge this and open a separate issue for refactoring this and moving the module. Is that ok?

**Pull request merged**